### PR TITLE
CLN: avoid values_from_object in Series

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #32422, #32419

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -7803,11 +7804,11 @@ def _reduce(
         self, op, name, axis=0, skipna=True, numeric_only=None, filter_type=None, **kwds
     ):
 
-        dtype_is_dt = self.dtypes.apply(lambda x: x.kind == "M")
+        dtype_is_dt = self.dtypes.apply(lambda x: x.kind == "M" or is_period_dtype(x))
```
Comment (jreback) [inline pandas/core/frame.py]: can you apply needs_i8_conversion instead here? (like you do below), this is a non-standard usage

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some comments

File: pandas/core/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/frame.py]: trouble is we don't want to include td64

Comment (jreback) [reply pandas/core/frame.py]: ok how about using is_datetime64_any_dtype then; this is a non-obvious pattern

File: pandas/core/frame.py [hunk 2]
```diff
@@ -7868,6 +7869,9 @@ def blk_func(values):
                 assert len(res) == max(list(res.keys())) + 1, res.keys()
             out = df._constructor_sliced(res, index=range(len(res)), dtype=out_dtype)
             out.index = df.columns
+            if axis == 0 and df.dtypes.apply(needs_i8_conversion).any():
```
Comment (jreback) [inline pandas/core/frame.py]: can you expand the comment to why this is needed here at all?

File: pandas/core/nanops.py [hunk 3]
```diff
@@ -281,23 +281,20 @@ def _get_values(
...
-        dtype = values.dtype
+    dtype = values.dtype
 
-    if is_datetime_or_timedelta_dtype(values) or is_datetime64tz_dtype(values):
+    if needs_i8_conversion(values):
         # changing timedelta64/datetime64 to int64 needs to happen after
         #  finding `mask` above
-        values = getattr(values, "asi8", values)
-        values = values.view(np.int64)
+        if isinstance(values, np.ndarray):
```
Comment (jreback) [inline pandas/core/nanops.py]: why is this case still here? when is this actually an ndarray? (or conversely, when is this a DTI/TDI). it is non-obvious how we get to this point.

File: pandas/core/nanops.py [hunk 4]
```diff
@@ -330,12 +328,14 @@ def _get_values(
 
 def _na_ok_dtype(dtype) -> bool:
     # TODO: what about datetime64tz?  PeriodDtype?
-    return not issubclass(dtype.type, (np.integer, np.timedelta64, np.datetime64))
+    if needs_i8_conversion(dtype):
```
Comment (jreback) [inline pandas/core/nanops.py]: comment is out-dated

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

File: pandas/tests/frame/test_analytics.py [hunk 5]
```diff
@@ -875,11 +875,6 @@ def test_mean_datetimelike(self):
         expected = pd.Series({"A": 1.0, "C": df.loc[1, "C"]})
         tm.assert_series_equal(result, expected)
 
-    @pytest.mark.xfail(
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: I guess you should technically have a whatsnew note as this 'bug' is fixed (do in followon)

**Comment by jreback** [comment]
thanks

File: pandas/core/series.py [hunk 6]
```diff
@@ -2055,7 +2055,7 @@ def idxmax(self, axis=0, skipna=True, *args, **kwargs):
         nan
         """
         skipna = nv.validate_argmax_with_skipna(skipna, args, kwargs)
-        i = nanops.nanargmax(com.values_from_object(self), skipna=skipna)
+        i = nanops.nanargmax(self._values, skipna=skipna)
```
Comment (jorisvandenbossche) [inline pandas/core/series.py]: nanargmax/nanargmin expect to get an ndarray. Due to this change, it is no longer guaranteed to be an ndarray. Reported this as https://github.com/pandas-dev/pandas/issues/32749

So those lines should either be reverted, or another "convert to ndarray" function should be used (or nanargmax/nanargmin could be rewritten to support EAs, but personally I think it is much cleaner to keep those algos based on numpy arrays)

Comment (jorisvandenbossche) [reply pandas/core/series.py:2058]: @jbrockmendel can you respond to this?

**Comment by jorisvandenbossche** [comment]
@jbrockmendel another problem with this PR is that you enabled "mean" for Period dtype (but only for DataFrames), while we had long discussions before (when initially adding mean support for datetimelikes) that ended in not supporting mean for period (-> https://github.com/pandas-dev/pandas/pull/24757/)

**Comment by jorisvandenbossche** [comment]
Opened a PR for that in the mean time: https://github.com/pandas-dev/pandas/pull/33758

File: pandas/core/series.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/series.py:2058]: > (or nanargmax/nanargmin could be rewritten to support EAs, but personally I think it is much cleaner to keep those algos based on numpy arrays)

I'd be fine with either of these options.  Probably prefer both actually: a EA-supporting public method and an ndarray-only private method for each of the relevant nanops funcs.

**Pull request merged**